examples/vision/siamese.py

```diff
@@ -0,0 +1,345 @@
+of contrasting [vector spaces](https://en.wikipedia.org/wiki/Vector_space)
+from which these sample inputs were taken.
+
+"""
+
+"""
+## Setup
+"""
+
+import random
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers.experimental.preprocessing import Rescaling
+from tensorflow.keras.layers import Flatten, Dense, Concatenate, Lambda, Input
+from tensorflow.keras.layers import Conv2D, Activation,AveragePooling2D
+from tensorflow.keras.datasets import mnist
+from tensorflow.keras import utils
+import matplotlib.pyplot as plt
+from keras import backend as K
+
+"""
+## Load the MNIST dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+
+# Change the data type to a floating point format
+x_train = x_train.astype('float32')
+x_test = x_test.astype('float32')
+
+"""
+## Create pairs of images
+
+We will train the model to differentiate each digit from one another. For
+example, digit `0` needs to be differentiated from the rest of the
+digits (`1` through `9`), digit `1` - from `0` and `2` through `9`, and so on.
+To carry this out, we will select N random images from class A (for example, for
+digit `0`) and pair it with N random images from another class B (for example,
+for digit `1`). Then, we can repeat this process for all classes of digits
+(until digit `9`). Once we have paired digit `0` with other digits, we can
+repeat this process for the remaining classes for the rest of the digits (from
+`1` until `9`).
+
+"""
+
+def make_pairs(x, y):
+    """
+    Parameters
+    ----------
+    x : list
+        List containing images, each index in this list corresponds to one image
+    y : list
+        List containing labels, each label with datatype of `int`
+
+    Returns
+    -------
+    Tuple containing two numpy arrays as (pair_of_samples, labels), 
+    where pair of samples' shape is (2len(x), 2,n_features_dims) and
+    labels are a binary array of shape (2len(x))
+    """
+
+    num_classes = max(y) + 1
+    digit_indices = [np.where(y == i)[0] for i in range(num_classes)]
+
+    pairs = []
+    labels = []
+
+    for idx1 in range(len(x)):
+        # add a matching example
+        x1 = x[idx1]
+        label1 = y[idx1]
+        idx2 = random.choice(digit_indices[label1])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [1]
+    
+        # add a non-matching example
+        label2 = random.randint(0, num_classes-1)
+        while label2 == label1:
+            label2 = random.randint(0, num_classes-1)
+
+        idx2 = random.choice(digit_indices[label2])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [0]
+
+    return np.array(pairs), np.array(labels).astype('float32')
+
+pairs_train, labels_train = make_pairs(x_train, y_train)
+pairs_test, labels_test = make_pairs(x_test, y_test)
+
+"""
+## Convert the data into TensorFlow Dataset objects
+<br>
```

Comment:
Don't use `br` tags, anywhere. Text sections in code examples are meant to be formatted as Markdown.